### PR TITLE
Implement staged warm-up adaptation schedule

### LINKED ISSUE #79: Allow user to define a schedule for adaptation during burn-in
We would like to allow for the user to specify a schedule that allows for different adaptation strategies to be used at different points during warm-up.

For example:

Stage 1: For the first 50 iterations, adapt the scale and use the identity matrix as a fixed proposal covariance.
Stage 2: For the next 50 iterations adapt the scale, and adaptively learn a diagonal proposal covariance.
Stage 3: For the remaining warm-up iterations adapt the scale and adaptively learn a dense proposal covariance.

At each stage adaptation parameters would be initialised at the values at the end of the previous stage.

This could be specified as an argument in the sample_chain function as
```
results <- sample_chain(
  # Other arguments
  adapters = list(
   list(scale_adaptor(), fixed_shape(), 50),
   list(scale_adaptor(), variance_shape_adapter(), 50),
   list(scale_adaptor(), covariance_shape_adapter())
)
)
```

In addition to the above, is it possible for the user to specify fixed scale and shape parameters (maybe this should be its own issue)?

### PR DESCRIPTION
**Pull request by jennajiali**
This PR implements the staged warm-up adaptation interface proposed in issue #79, allowing users to specify different adapter configurations at different points during warm-up. It builds on the partial implementation started on the mmg/adapter-staging branch and ensures adapter states correctly carry over across stage boundaries.

## Background
Previously, `sample_chain()` ran all adapters from iteration 1 of warm-up with no way to vary which adapters were active over time. This is suboptimal in practice: the shape adapter (which estimates the proposal covariance) should not start until the scale adapter has had enough iterations to stabilise, because early covariance estimates based on poorly-mixed samples can make the sampler worse rather than better.

Furthermore, previously all adapter `initialize()` functions ignored the current proposal parameters on re-initialisation, causing the scale and shape estimates accumulated during one staged warm-up stage to be discarded at the start of the next. This contradicted the requirement in issue #79 that adaptation parameters be initialised at the values at the end of the previous stage.

### R/chains.R
* `check_and_process_adapters()` — new internal function that normalises the `adapters` argument into a canonical list of stage specifications, each with `$adapters` (a list of adapter objects) and `$n_iteration` (an integer). Accepts three input forms:
    1.  A flat list of adapter objects (existing behaviour, now wrapped into a single stage)
    2.  A staged list where each element is a list of adapters with an optional trailing integer iteration count; the last stage may omit the count and receives the remaining warm-up iterations
    3.  A function accepting `n_warm_up_iteration` and returning a staged list (used by `progressive_adaptation_schedule()`)
* `sample_chain()` — warm-up loop now iterates over stages produced by `check_and_process_adapters()`, passing each stage's adapter list and iteration count to `chain_loop()` separately. Several bugs in the initial partial implementation are fixed:
    * Bootstrap `warm_up_results` initialised with key `$final_state` (not `$state`) to be consistent with `chain_loop()`'s return value
    * Each stage now reads `stage$n_iteration` rather than always using the full `n_warm_up_iteration`
    * Each stage now passes `stage$adapters` (clean list of adapter objects) rather than the whole named stage list
    * Progress bar label now shows current stage number (e.g. `Warm-up (stage 1/3)`)
* `combine_warm_up_results()` — rewritten to use `rbind` instead of `mapply(c, ...)`. This correctly handles the bootstrap case (`rbind(NULL, matrix)` returns the matrix unchanged) and uses the correct `$final_state` key throughout
* `combine_stage_results()` — fixed `$state` → `$final_state` to match `chain_loop()`'s return value
* `@param adapters` documentation — updated to describe all three input forms

### R/adaptation.R
* `progressive_adaptation_schedule()` — new exported convenience function that returns a schedule constructor. When called with `n_warm_up_iteration`, it produces a three-stage adaptation schedule following common practice:
    * Stage 1: scale adapter only (shape held fixed at identity), for `n_fixed_shape_iteration` iterations (default 50)
    * Stage 2: scale adapter + diagonal shape adapter (`shape_adapter("variance")`), for `n_diagonal_shape_iteration` iterations (default 50)
    * Stage 3: scale adapter + dense shape adapter (`shape_adapter("covariance")`), for remaining iterations
    * If the sum of stage 1 and stage 2 counts exceeds `n_warm_up_iteration`, both are reduced proportionally and stage 3 is skipped. All adapter objects and iteration counts are customisable via arguments. Includes full roxygen documentation and a working example.
* `stochastic_approximation_scale_adapter` and `dual_averaging_scale_adapter`: when `initial_scale` is not user-specified, `initialize()` now reads `proposal$parameters()$scale` and uses it as the starting log-scale if available, falling back to `proposal$default_initial_scale()` only when no current scale exists. Note that for `dual_averaging_scale_adapter`, `mu`, `smoothed_log_scale` and `accept_prob_error` already persisted across stages via closure scope and required no change.
* `variance_shape_adapter` and `covariance_shape_adapter`: `initialize()` now follows a three-priority order:
    1.  Explicit `initial_shape` constructor argument if supplied.
    2.  Current proposal shape, guarded by type compatibility checks. `variance_shape_adapter` requires a vector of matching length to prevent accidentally carrying over a matrix-valued shape. `covariance_shape_adapter` requires a matrix of matching dimensions (Cholesky factor) to prevent carrying over a vector-valued shape.
    3.  Default fallback (unit variances or identity matrix, respectively).
* Added `initial_shape` parameter (numeric vector of per-dimension scales for variance; lower-triangular Cholesky factor matrix for covariance) to the respective function signatures and roxygen documentation, addressing the request to allow users to supply domain-knowledge-based starting values.

### tests/testthat/test-chains.R
New tests appended after existing tests:
* Unit tests for `check_and_process_adapters()` covering all three input forms, the last-stage remainder logic, and all error cases (non-last stage missing count, counts not summing to `n_warm_up_iteration`, invalid input types)
* Integration tests for `sample_chain()` with staged adapters: correct warm-up row counts across two and three stages, correct behaviour with differing adapter sets across stages when `trace_warm_up = FALSE`, function-form adapters argument, correct `final_state` type, and invalid adapters error propagation

### tests/testthat/test-adaptation.R
New tests appended after existing tests covering `progressive_adaptation_schedule()`:
* Returns a function
* Correct stage count and iteration counts for the normal case (`n_warm_up` > `n_fixed` + `n_diagonal`), the exact-boundary case (no dense stage), and the fallback case (`n_warm_up` < `n_fixed` + `n_diagonal`)
* Edge case of `n_warm_up_iteration = 1`
* Custom `n_fixed_shape_iteration` and `n_diagonal_shape_iteration` are respected
* Correct adapter count in each stage
* Schedule output always parses cleanly through `check_and_process_adapters()` across a range of `n_warm_up_iteration` values
* End-to-end `sample_chain()` integration test

New tests covering state carry-over across stages:
* Four tests for scale adapter carry-over: both `stochastic_approximation` and `dual_averaging` variants are checked for (a) reading the current proposal scale on re-initialisation and (b) explicit `initial_scale` still taking priority over the proposal's current value.
* Four tests for `variance_shape_adapter`: carry-over from a vector proposal shape; explicit `initial_shape` override; `NULL` proposal falls back to unit variances; matrix-valued proposal shape (incompatible type) falls back to unit variances.
* Four tests for `covariance_shape_adapter`: carry-over from a matrix proposal shape; explicit `initial_shape` override; `NULL` proposal falls back to identity; vector-valued proposal shape (incompatible type) falls back to identity.
* One end-to-end integration test via `sample_chain` with `trace_warm_up=TRUE`: verifies that the `log_scale` in `warm_up_statistics` does not jump back to the default value at the stage boundary between two consecutive scale-adapter-only stages.

## Usage examples

Fully custom staged schedule:
```r
results <- sample_chain(
  target_distribution,
  initial_state = rnorm(2),
  n_warm_up_iteration = 1000,
  n_main_iteration = 1000,
  adapters = list(
    list(scale_adapter(), 50),                              # scale only
    list(scale_adapter(), shape_adapter("variance"), 50),   # scale + diagonal shape
    list(scale_adapter(), shape_adapter("covariance"))      # scale + dense shape (remainder)
  )
)
```

Using the convenience constructor with defaults:
```r
results <- sample_chain(
  target_distribution,
  initial_state = rnorm(2),
  n_warm_up_iteration = 1000,
  n_main_iteration = 1000,
  adapters = progressive_adaptation_schedule()
)
```

Existing flat-list usage is unchanged:
```r
results <- sample_chain(
  target_distribution,
  initial_state = rnorm(2),
  n_warm_up_iteration = 1000,
  n_main_iteration = 1000,
  adapters = list(scale_adapter(), shape_adapter())  # still works as before
)
```

## Known limitation
When trace_warm_up = TRUE and stages use different adapter sets (e.g. scale-only in stage 1, scale + shape in stage 2), the warm-up statistics matrices from each stage have different column counts and cannot be rbind-ed. This causes an error. A future PR could address this by filling NA for missing columns. For now, trace_warm_up = TRUE with staged adapters is only safe when all stages use the same adapter set. This limitation is documented in the new tests.

### CODE REVIEW

File: R/chains.R [hunk 1]
```diff
@@ -167,6 +190,77 @@ check_and_process_target_distribution <- function(target_distribution) {
...
+      stage_adapters <- stage_spec[-length(stage_spec)]
+    } else {
+      # No iteration count supplied — only valid for the last stage
+      if (i < length(adapters)) {
+        stop(paste(
+          "Only the last stage may omit an iteration count.",
+          sprintf("Stage %d has no iteration count.", i)
+        ))
+      }
+      stage_n_iter <- n_warm_up_iteration - n_allocated
```
Comment (matt-graham) [inline R/chains.R:272]:
```suggestion
      stage_n_iter <- n_warm_up_iteration - n_allocated
      if (state_n_iter < 0) stop("Per-stage iteration counts exceeds n_warm_up_iteration")
```
I think there is nothing to stop the user up to this point specifying iteration counts per stage that exceed `n_warm_up_iteration` which would make `stage_n_iter` negative here which would result in an error in the `seq_len` call in `chain_loop`, so it is probably better to give a more interpretable message here.

File: R/chains.R [hunk 2]
```diff
@@ -285,9 +379,22 @@ chain_loop <- function(
   list(final_state = state, traces = traces, statistics = statistics)
 }
 
+combine_warm_up_results <- function(warm_up_results_1, warm_up_results_2) {
+  # warm_up_results_1$traces and $statistics are NULL either because this is
```
Comment (matt-graham) [inline R/chains.R]:
```suggestion
  # warm_up_results_1$traces and $statistics may be NULL either because this is
```

I think this comment is useful but as far as I can tell `warm_up_results_1$traces` and `warm_up_results_1$statistics` are not guaranteed to be `NULL` its just that may be.

File: R/adaptation.R [hunk 3]
```diff
@@ -77,7 +78,14 @@ stochastic_approximation_scale_adapter <- function(
...
     if (is.null(initial_scale)) {
-      initial_scale <- proposal$default_initial_scale(initial_state$dimension())
+      # Prefer the current proposal scale (e.g. carried over from a previous
+      # warm-up stage) over the proposal/dimension-dependent default.
+      current_scale <- proposal$parameters()$scale
+      initial_scale <- if (!is.null(current_scale)) {
+        current_scale
+      } else {
+        proposal$default_initial_scale(initial_state$dimension())
+      }
```
Comment (matt-graham) [inline R/adaptation.R]: As this code snippet is duplicated in dual-averaging scale adapter below it would be good to factor it out in to a function to be reused at both points e.g. something like

```R
get_initial_scale <- function(proposal, initial_state) {
    # Prefer the current proposal scale (e.g. carried over from a previous
    # warm-up stage) over the proposal/dimension-dependent default.
    current_scale <- proposal$parameters()$scale
    if (!is.null(current_scale)) {
      current_scale
    } else {
      proposal$default_initial_scale(initial_state$dimension())
    }
}
```

File: R/adaptation.R [hunk 4]
```diff
@@ -237,12 +259,30 @@ shape_adapter <- function(type = "covariance", kappa = 1) {
...
+    } else {
+      # Priority 2: current proposal shape, if it is a vector of the right
+      # length (i.e. from a previous variance_shape_adapter stage).
+      # The length(current_shape) == dim guard is important:
+      # if the previous stage used a covariance_shape_adapter, the
+      # proposal's shape is a matrix, not a vector.
+      # Squaring a matrix is meaningless here so we fall back to identity.
+      # Priority 3: fall back to unit variances.
+      current_shape <- proposal$parameters()$shape
+      variance_estimate <<- if (!is.null(current_shape) && length(current_shape) == dim) {
```
Comment (matt-graham) [inline R/adaptation.R]:
```suggestion
      current_shape_is_vector = length(current_shape) == dim
      variance_estimate <<- if (!is.null(current_shape) && current_shape_is_vector) {
```
This will both make code a bit more self-documenting and should solve linter line length warning

File: R/adaptation.R [hunk 5]
```diff
@@ -291,13 +339,29 @@ variance_shape_adapter <- function(kappa = 1) {
...
+      # Priority 1: explicit user-supplied starting Cholesky factor.
+      chol_covariance_estimate <<- initial_shape
+    } else {
+      # Priority 2: current proposal shape, if it is a square matrix of the
+      # right dimension (i.e. from a previous shape adapter stage).
+      # Priority 3: fall back to identity matrix.
+      current_shape <- proposal$parameters()$shape
+      chol_covariance_estimate <<- if (!is.null(current_shape) &&
+        is.matrix(current_shape) &&
+        nrow(current_shape) == dim) {
```
Comment (matt-graham) [inline R/adaptation.R]:
```suggestion
      current_shape_valid <- is.matrix(current_shape)  && nrow(current_shape) == dim)
      chol_covariance_estimate <<- if (!is.null(current_shape) && current_shape_valid) {
```
Factoring out condition here into variable again makes code a bit more readable and may help with line length warning (though haven't counted to check!)

**Review by matt-graham — CHANGES_REQUESTED** [review]
This is looke great @jennajiali - I've added some suggested changes but mainly minor things. Will also need to resolve the linter warnings

File: R/chains.R [hunk 6]
```diff
@@ -174,6 +197,78 @@ check_and_process_target_distribution <- function(target_distribution) {
...
+    "adapters invalid - must be a flat list of adapter objects, a list of",
+    "stage specifications (each a list of adapters with an optional trailing",
+    "integer iteration count), or a function accepting n_warm_up_iteration."
+  )
+  if (is.function(adapters)) {
+    # Form 3: schedule constructor function
+    stages <- adapters(n_warm_up_iteration)
+    return(check_and_process_adapters(stages, n_warm_up_iteration))
+  }
+  if (!is.list(adapters)) stop(error_message)
```
Comment (matt-graham) [inline R/chains.R]: This case is currently not covered by a test - i.e. that we stop with an error message if the user passes an `adapters` argument which isn't of the expected format. It might be worth adding a quick test with [the `testthat` `expect_error` function](https://testthat.r-lib.org/reference/expect_error.html) to check giving invalid input raises an error as expected.

File: R/chains.R [hunk 7]
```diff
@@ -174,6 +197,78 @@ check_and_process_target_distribution <- function(target_distribution) {
...
+    } else {
+      # No iteration count supplied — only valid for the last stage
+      if (i < length(adapters)) {
+        stop(paste(
+          "Only the last stage may omit an iteration count.",
+          sprintf("Stage %d has no iteration count.", i)
+        ))
+      }
+      stage_n_iter <- n_warm_up_iteration - n_allocated
+      if (stage_n_iter < 0) stop("Per-stage iteration counts exceeds n_warm_up_iteration")
```
Comment (matt-graham) [inline R/chains.R]:
```suggestion
      if (stage_n_iter < 0) {
        stop("Per-stage iteration counts exceeds n_warm_up_iteration")
      }
```

This should resolve linter line length warning here

Comment (matt-graham) [reply R/chains.R]: Also similar to above this case is currently not covered by a test - it would be good to add a test with `expect_error` that checks if we specify adaptation schedule with more number of iterations than `n_warm_up_iteration` that we do get this informative error (as opposed to failing later when attempting to sample with a negative number of stage iterations).

File: R/chains.R [hunk 8]
```diff
@@ -174,6 +197,78 @@ check_and_process_target_distribution <- function(target_distribution) {
...
+  if (n_allocated != n_warm_up_iteration) {
+    stop(sprintf(
+      paste(
+        "Sum of stage iteration counts (%d) does not equal",
+        "n_warm_up_iteration (%d)."
+      ),
+      n_allocated, n_warm_up_iteration
+    ))
+  }
+  stages
```
Comment (matt-graham) [inline R/chains.R:291]: This part of function is relatively self-contained in terms of the operations it is performing and inputs required so it may be worth factoring it out in to a separate function, for example something like
```R
process_stage_specifications <- function(
    adapters, n_warm_up_iteration, error_message
) {
  stages <- vector("list", length(adapters))
  ...  # Rest of existing implementation goes here
  stages
}
```

so that existing `check_and_process_adapters` becomes something like

```R
check_and_process_adapters <- function(adapters, n_warm_up_iteration) {
  error_message <- paste(
    "adapters invalid - must be a flat list of adapter objects, a list of",
    "stage specifications (each a list of adapters with an optional trailing",
    "integer iteration count), or a function accepting n_warm_up_iteration."
  )
  adapter_stages <- if (all(sapply(adapters, is_adapter))) {
    # Form 1: flat list of adapter objects
    list(list(adapters = adapters, n_iteration = n_warm_up_iteration))
  } else if (all(sapply(adapters, is.list))) {
    # Form 2: list of stage specifications
    adapters
  } else if (is.function(adapters)) {
    # Form 3: schedule constructor function
    adapters(n_warm_up_iteration)
  } else {
    stop(error_message)
  }
  process_stage_specifications(
      adapter_stages, n_warm_up_iteration, error_message
  )
}
```

to make top-level logic corresponding to the three different accepted forms more obvious and reduce the complexity of the function (which is what is triggering the linter warning).

**Review by matt-graham — CHANGES_REQUESTED** [review]
Thanks @jennajiali for addressing previous comments. I've added a few more small suggestions to resolve the linter and test coverage fails in the checks. Let me know if anything doesn't make sense.

File: R/chains.R (hunk 8 above)
Comment (jennajiali) [reply R/chains.R:291]: Thanks Matt. I've pushed a fix for this in 72ffc534da1e93fb585c6eb791973a26fc4dbbfa, building on your suggested refactoring. Let me know if anything is unclear or needs further changes.

File: tests/testthat/test-chains.R [hunk 9]
```diff
@@ -229,6 +229,74 @@ test_that(
...
+    expect_length(stages[[1]]$adapters, 1L)
+    expect_equal(stages[[2]]$n_iteration, 50L)
+    expect_length(stages[[2]]$adapters, 0L)
+    expect_equal(stages[[3]]$n_iteration, 20L)
+    expect_length(stages[[3]]$adapters, 1L)
+  }
+)
+
+test_that(
+  "check_and_process_adapters: empty top-level list is treated as a single no-adapter stage covering all warm-up iterations",
```
Comment (matt-graham) [inline tests/testthat/test-chains.R:288]: Great idea to have this as an additional test case!

File: tests/testthat/test-chains.R [hunk 10]
```diff
@@ -246,6 +314,32 @@ test_that(
 
 # ── sample_chain with staged adapters integration tests ───────────────────────
 
+test_that(
+  "sample_chain with an empty middle stage (no adapters, just an iteration count) runs without error",
```
Comment (matt-graham) [inline tests/testthat/test-chains.R:318]: This is also a great test to have - it's a testament to the generality of your implementation that this works as expected!

**Review by matt-graham — APPROVED** [review]
Thanks @jennajiali for making the additional changes. Factoring out the `process_stage_specifications` function I think has helped improve the readability a lot - well done for also catching and fixing the unintentional bug in my suggested changes in the handling of the form 1 case 😅. The additional tests for edge cases are also great. I merged in the changes from #130 and made one small fix to a use of `=` rather than `<-` that was causing styler check in `pre-commit` to fail. This all looks good to go to me now so will merge in.
